common/src/main/java/com/microsoft/identity/common/internal/controllers/NativeAuthMsalController.kt

```diff
@@ -39,17 +86,735 @@ class NativeAuthMsalController : BaseNativeAuthController() {
+                        details = result.details,
+                        errorCodes = result.errorCodes
+                    )
+                }
+            }
+        } catch (e: Exception) {
+            Logger.error(TAG, "Exception thrown in signInResendCode", e)
+            throw e
+        }
+    }
+
+    /**
+     * Makes a call to the submit password endpoint, mapping responses returned from the server into a command result.
+     */
+    fun signInSubmitPassword(parameters: SignInSubmitPasswordCommandParameters): SignInSubmitPasswordCommandResult {
+        LogSession.logMethodCall(TAG, "${TAG}.signInSubmitPassword")
+
+        try {
+            val oAuth2Strategy = createOAuth2Strategy(parameters)
+
+            val mergedScopes = addDefaultScopes(parameters.scopes)
+            val parametersWithScopes =
+                CommandUtil.createSignInSubmitPasswordCommandParametersWithScopes(
+                    parameters,
+                    mergedScopes
+                )
+
+            return performPasswordTokenCall(
+                oAuth2Strategy = oAuth2Strategy,
+                parameters = parametersWithScopes
+            ).toSignInSubmitPasswordCommandResult(
+                oAuth2Strategy = oAuth2Strategy,
+                parametersWithScopes = parametersWithScopes
+            )
+        } catch (e: Exception) {
+            Logger.error(TAG, "Exception thrown in signInSubmitPassword", e)
+            throw e
+        }
+    }
+
+    /**
+     * Native-auth specific implementation of fetching a token from the cache, and/or refreshing it.
+     * Main differences with standard implementation in [LocalMSALController] are:
+     * - No scopes are passed in as part of the (developer provided) parameters. The scopes from the
+     * AT are used to make the refresh token call.
+     * - When the RT is expired or the refresh token call fails, a different exception is thrown.
+     */
+    @Throws(
+        IOException::class,
+        ClientException::class,
+        ArgumentException::class,
+        ServiceException::class
+    )
     fun acquireTokenSilent(
         parameters: AcquireTokenNoFixedScopesCommandParameters
     ): AcquireTokenResult {
         LogSession.logMethodCall(TAG, "${TAG}.acquireTokenSilent")
 
         val acquireTokenSilentResult = AcquireTokenResult()
-        
+
+        // Validate original AcquireTokenNoScopesCommandParameters parameters
+        parameters.validate()
+
+        // Convert AcquireTokenNoScopesCommandParameters into SilentTokenCommandParameters
+        // so we can re-use existing MSAL logic
+        val silentTokenCommandParameters =
+            CommandUtil.convertAcquireTokenNoFixedScopesCommandParameters(
+                parameters
+            )
+        // Not adding any (default) scopes, because we want to retrieve all tokens (regardless of scope).
+        // Scopes will be added later in the flow, if necessary.
```

Comment:
Can you elaborate on this?

Reply:
@shahzaibj  we've been over this during the technical design discussions. What further would you like us to explain here?

Reply:
I think just explain in the comments here what was already explained in the doc. Elaborate more on what we mean by added "later" and "if necessary"

Reply:
Done